### ISSUE TITLE
torch.bmm in FP16 mode, result is different from loop through batch dimension and call torch.mm

### ISSUE THREAD

**Issue by wezuo** [pytorch/pytorch#59331]
## 🐛 Bug

When run torch.bmm in FP16 mode, compare the result with a loop through the batch dimension of the input, and run torch.mm in each iteration. The results do not match. 

## To Reproduce

1. run the following code sample
1. The result:

```  
        assert torch.allclose(res, res_half)
    AssertionError
```

The code sample:
```
import torch

torch.manual_seed(1234)

mat_a = torch.rand(512, 8, 64).cuda()
mat_b = torch.rand(512, 64, 8).cuda()

mat_a_half = mat_a.half()
mat_b_half = mat_b.half()

res_list = []
for i in range(0, 512):
  res_tmp =torch.mm(mat_a_half[i], mat_b_half[i])
  res_list += [res_tmp]
res = torch.cat(res_list, dim=0).reshape(512,8,8)

res_half = torch.bmm(mat_a_half, mat_b_half)

assert torch.allclose(res, res_half)                                                 
```
When I tried smaller batch size (in the sample is 512, if I try for example 256, 100, etc) the assertion succeeds.

## Expected behavior
The assertion should not fail, the result of both torch.bmm and a loop through torch.mm should be identical.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```

# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py

CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.19.6
Libc version: glibc-2.10

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration:
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB
GPU 6: Tesla V100-SXM2-32GB
GPU 7: Tesla V100-SXM2-32GB

Nvidia driver version: 450.80.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A
```
## Additional context

cc @ngimel

**Comment by ngimel** [comment]
Result is well within expected accuracy for half, the difference is ~0.01 for values that are O(10). Batched matmul is not guaranteed to return identical results as a loop over matrices
```
In [3]: import torch
   ...: 
   ...: torch.manual_seed(1234)
   ...: 
   ...: mat_a = torch.rand(512, 8, 64).cuda()
   ...: mat_b = torch.rand(512, 64, 8).cuda()
   ...: 
   ...: mat_a_half = mat_a.half()
   ...: mat_b_half = mat_b.half()
   ...: 
   ...: res_list = []
   ...: for i in range(0, 512):
   ...:   res_tmp =torch.mm(mat_a_half[i], mat_b_half[i])
   ...:   res_list += [res_tmp]
   ...: res = torch.cat(res_list, dim=0).reshape(512,8,8)
   ...: 
   ...: res_half = torch.bmm(mat_a_half, mat_b_half)
   ...: 
   ...: for i in range(512):
   ...:     if (res_half[i]-res[i]).abs().max()>0.01:
   ...:         print("i: ", i, res_half[i], res[i])
   ...: 
   ...: 
   ...: 
i:  424 tensor([[12.7031, 13.2812, 17.0625, 17.1094, 13.7734, 13.4375, 15.4375, 12.1406],
        [15.0703, 17.7500, 19.2969, 20.4844, 16.8750, 16.0312, 17.0938, 15.1406],
        [13.7812, 15.2891, 19.7188, 20.4062, 15.6719, 15.5234, 17.1250, 13.7734],
        [13.6797, 14.0469, 16.7969, 17.7969, 16.9531, 13.6875, 15.2578, 12.5078],
        [13.9609, 14.8047, 16.6406, 18.0312, 16.3906, 13.9609, 15.1641, 12.0938],
        [13.5312, 16.0469, 17.9531, 19.4062, 17.3906, 14.5469, 17.2969, 15.5234],
        [14.4766, 16.6719, 19.4375, 19.6094, 18.2344, 15.2734, 16.5781, 14.8125],
        [13.4453, 14.1641, 16.2656, 19.1406, 16.2969, 14.7656, 15.8828, 13.6016]],
       device='cuda:0', dtype=torch.float16) tensor([[12.7031, 13.2812, 17.0625, 17.1094, 13.7734, 13.4375, 15.4375, 12.1406],
        [15.0703, 17.7500, 19.2969, 20.4844, 16.8750, 16.0312, 17.0938, 15.1406],
        [13.7812, 15.2891, 19.7188, 20.4062, 15.6719, 15.5234, 17.1250, 13.7734],
        [13.6797, 14.0469, 16.7969, 17.7969, 16.9531, 13.6875, 15.2578, 12.5078],
        [13.9609, 14.8047, 16.6406, 18.0312, 16.3906, 13.9609, 15.1641, 12.0938],
        [13.5312, 16.0469, 17.9531, 19.3906, 17.3906, 14.5469, 17.2969, 15.5234],
        [14.4766, 16.6719, 19.4375, 19.6094, 18.2344, 15.2734, 16.5781, 14.8125],
        [13.4453, 14.1641, 16.2656, 19.1406, 16.2969, 14.7656, 15.8828, 13.6016]],
       device='cuda:0', dtype=torch.float16)
```
